### PR TITLE
Multiple inputs and outputs in the python operator

### PR DESCRIPTION
**Pull request by banasraf**
* Added multiple inputs and outputs in the python operator
* Optimized passing input to a called function. Now data is not copied but shared by numpy arrays.
* Moved some utility functions from backend_impl to a header so I could reuse them in the operator.
* Updated the jupyter notebook to show a usage of an operator with more than one input and output.

### CODE REVIEW

**Comment by banasraf** [comment]
!build

File: dali/pipeline/operators/python_function/python_function.cc [hunk 1]
```diff
@@ -115,17 +66,40 @@ void CopyNumpyArrayToTensor(Tensor<CPUBackend> &tensor, const py::array &array)
...
-    py::array output_array = python_function(TensorToNumpyArray(input));
-    CopyNumpyArrayToTensor(output, output_array);
+    output_o = python_function(*py::tuple(args_list));
   } catch(const py::error_already_set & e) {
     throw std::runtime_error(to_string("PythonFunction error: ") + to_string(e.what()));
   }
+  py::tuple output = (py::tuple::check_(output_o)) ? output_o : py::make_tuple(output_o);
+  DALI_ENFORCE(output.size() == static_cast<size_t>(ws->NumOutput()),
+               "Python function returned " + std::to_string(output.size()) + " outputs and "
+                   + std::to_string(ws->NumOutput()) + " were expected.");
```
Comment (awolant) [inline dali/pipeline/operators/python_function/python_function.cc]: Maybe add test for this in `test_python_function_operator.py` ?

Comment (banasraf) [reply dali/pipeline/operators/python_function/python_function.cc]: Added.

**Comment by banasraf** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by banasraf** [comment]
!build

**Pull request merged**